File: objectiveai-js/src/functions/profiles/computations/http.ts
import { ObjectiveAI, RequestOptions } from "../../../client";
import type { Remote } from "../../remote";
import { Stream } from "../../../stream";
import { FunctionProfileComputationChunk } from "./response/streaming/function_profile_computation_chunk";
import { FunctionProfileComputation } from "./response/unary/function_profile_computation";
import {
  FunctionProfileComputationCreateParamsInlineFunction,
  FunctionProfileComputationCreateParamsInlineFunctionNonStreaming,
  FunctionProfileComputationCreateParamsInlineFunctionStreaming,
  FunctionProfileComputationCreateParamsRemoteFunction,
  FunctionProfileComputationCreateParamsRemoteFunctionNonStreaming,
  FunctionProfileComputationCreateParamsRemoteFunctionStreaming,
} from "./request/function_profile_computation_create_params";
import { InlineFunction } from "../../function";

export function inlineFunctionCreate(
  client: ObjectiveAI,
  body: FunctionProfileComputationCreateParamsInlineFunctionStreaming,
  options?: RequestOptions,
): Promise<Stream<FunctionProfileComputationChunk>>;
export function inlineFunctionCreate(
  client: ObjectiveAI,
  body: FunctionProfileComputationCreateParamsInlineFunctionNonStreaming,
  options?: RequestOptions,
): Promise<FunctionProfileComputation>;
export function inlineFunctionCreate(
  client: ObjectiveAI,
  body: FunctionProfileComputationCreateParamsInlineFunction,
  options?: RequestOptions,
): Promise<
  Stream<FunctionProfileComputationChunk> | FunctionProfileComputation
> {
  if (body.stream) {
    return client.post_streaming<FunctionProfileComputationChunk>(
      "/functions/profiles/compute",
      body,
      options,
    );
  }
  return client.post_unary<FunctionProfileComputation>(
    "/functions/profiles/compute",
    body,
    options,
  );
}

export function remoteFunctionCreate(
  client: ObjectiveAI,
  fremote: Remote,
  fowner: string,
  frepository: string,
  fcommit: string | null | undefined,
  body: FunctionProfileComputationCreateParamsRemoteFunctionStreaming,
  options?: RequestOptions,
): Promise<Stream<FunctionProfileComputationChunk>>;
export function remoteFunctionCreate(
  client: ObjectiveAI,
  fremote: Remote,
  fowner: string,
  frepository: string,
  fcommit: string | null | undefined,
  body: FunctionProfileComputationCreateParamsRemoteFunctionNonStreaming,
  options?: RequestOptions,
): Promise<FunctionProfileComputation>;
export function remoteFunctionCreate(
  client: ObjectiveAI,
  fremote: Remote,
  fowner: string,
  frepository: string,
  fcommit: string | null | undefined,
  body: FunctionProfileComputationCreateParamsRemoteFunction,
  options?: RequestOptions,
): Promise<
  Stream<FunctionProfileComputationChunk> | FunctionProfileComputation
> {
  const path =
    fcommit !== null && fcommit !== undefined
      ? `/functions/${fremote}/${fowner}/${frepository}/${fcommit}/profiles/compute`
      : `/functions/${fremote}/${fowner}/${frepository}/profiles/compute`;
  if (body.stream) {
    return client.post_streaming<FunctionProfileComputationChunk>(
      path,
      body,
      options,
    );
  }
  return client.post_unary<FunctionProfileComputation>(path, body, options);
}

export function create(
  client: ObjectiveAI,
  function_:
    | InlineFunction
    | {
        remote: Remote;
        owner: string;
        repository: string;
        commit?: string | null | undefined;
      },
  body: FunctionProfileComputationCreateParamsInlineFunctionStreaming,
  options?: RequestOptions,
): Promise<Stream<FunctionProfileComputationChunk>>;
export function create(
  client: ObjectiveAI,
  function_:
    | InlineFunction
    | {
        remote: Remote;
        owner: string;
        repository: string;
        commit?: string | null | undefined;
      },
  body: FunctionProfileComputationCreateParamsInlineFunctionNonStreaming,
  options?: RequestOptions,
): Promise<FunctionProfileComputation>;
export function create(
  client: ObjectiveAI,
  function_:
    | InlineFunction
    | {
        remote: Remote;
        owner: string;
        repository: string;
        commit?: string | null | undefined;
      },
  body: FunctionProfileComputationCreateParamsInlineFunction,
  options?: RequestOptions,
): Promise<
  Stream<FunctionProfileComputationChunk> | FunctionProfileComputation
> {
  if ("owner" in function_ && "repository" in function_) {
    const requestBody: FunctionProfileComputationCreateParamsRemoteFunction =
      body;
    if (requestBody.stream) {
      return remoteFunctionCreate(
        client,
        function_.remote,
        function_.owner,
        function_.repository,
        function_.commit ?? null,
        body as FunctionProfileComputationCreateParamsRemoteFunctionStreaming,
        options,
      );
    } else {
      return remoteFunctionCreate(
        client,
        function_.remote,
        function_.owner,
        function_.repository,
        function_.commit ?? null,
        body as FunctionProfileComputationCreateParamsRemoteFunctionNonStreaming,
        options,
      );
    }
  } else {
    const requestBody: FunctionProfileComputationCreateParamsInlineFunction = {
      ...body,
      function: function_,
    };
    if (requestBody.stream) {
      return inlineFunctionCreate(
        client,
        requestBody as FunctionProfileComputationCreateParamsInlineFunctionStreaming,
        options,
      );
    } else {
      return inlineFunctionCreate(
        client,
        requestBody as FunctionProfileComputationCreateParamsInlineFunctionNonStreaming,
        options,
      );
    }
  }
}

File: objectiveai-rs/src/functions/profiles/computations/http.rs
use crate::{HttpClient, HttpError};
use futures::Stream;

pub async fn compute_profile_unary(
    client: &HttpClient,
    request: super::request::Request,
) -> Result<super::response::unary::FunctionProfileComputation, HttpError> {
    match request {
        super::request::Request::FunctionInline { mut body } => {
            body.base.stream = None;
            client
                .send_unary(
                    reqwest::Method::POST,
                    "functions/profiles/compute",
                    Some(body),
                )
                .await
        }
        super::request::Request::FunctionRemote { path, mut body } => {
            body.stream = None;
            client
                .send_unary(
                    reqwest::Method::POST,
                    match path.fcommit {
                        Some(fcommit) => {
                            format!(
                                "functions/{}/{}/{}/profiles/compute",
                                path.fowner, path.frepository, fcommit
                            )
                        }
                        None => format!(
                            "functions/{}/{}/profiles/compute",
                            path.fowner, path.frepository
                        ),
                    },
                    Some(body),
                )
                .await
        }
    }
}

pub async fn compute_profile_streaming(
    client: &HttpClient,
    request: super::request::Request,
) -> Result<
    impl Stream<
        Item = Result<
            super::response::streaming::FunctionProfileComputationChunk,
            HttpError,
        >,
    > + Send
    + 'static,
    HttpError,
> {
    match request {
        super::request::Request::FunctionInline { mut body } => {
            body.base.stream = Some(true);
            Ok(futures::future::Either::Left(
                client
                    .send_streaming(
                        reqwest::Method::POST,
                        "functions/profiles/compute",
                        Some(body),
                    )
                    .await?,
            ))
        }
        super::request::Request::FunctionRemote { path, mut body } => {
            body.stream = Some(true);
            Ok(futures::future::Either::Right(
                client
                    .send_streaming(
                        reqwest::Method::POST,
                        match path.fcommit {
                            Some(fcommit) => {
                                format!(
                                    "functions/{}/{}/{}/profiles/compute",
                                    path.fowner, path.frepository, fcommit
                                )
                            }
                            None => format!(
                                "functions/{}/{}/profiles/compute",
                                path.fowner, path.frepository
                            ),
                        },
                        Some(body),
                    )
                    .await?,
            ))
        }
    }
}
